src/osd/OSDMap.h

```diff
@@ -833,21 +835,33 @@ class OSDMap {
   }
 
   unsigned get_crush_node_flags(int osd) const;
+  unsigned get_device_class_flags(int id) const;
 
+  bool _has_osd_state_flag(int osd, unsigned state) const {
+    if (test_flag(state))
+      return true; // global?
+    if (!exists(osd))
+      return false;
+    if (osd_state[osd] & state)
+      return true; // per osd?
+    if (get_crush_node_flags(osd) & state)
+      return true; // crush node?
+    auto class_id = crush->get_item_class_id(osd);
+    if (class_id >= 0 && (get_device_class_flags(class_id) & state))
+      return true; // device class?
+    return false;
+  }
   bool is_noup(int osd) const {
```

Comment:
I didn't change these before because of the code here:
```
        case OP_NOIN:
          if (osdmap.is_noin(osd)) {
            if (pending_inc.pending_osd_state_clear(osd, CEPH_OSD_NOIN))
              any = true;
          } else {
            pending_inc.pending_osd_state_set(osd, CEPH_OSD_NOIN);
            any = true;
          }
```
That needs to be updated to look at the explicit per-osd flag instead of using this helper if we're going to change the helper.

Reply:
(please git grep to check for any other callers that need updating too)